src/networking.c

```diff
@@ -4332,44 +4333,44 @@ void killIOThreads(void) {
     }
 }
 
-void startThreadedIO(void) {
-    serverAssert(server.io_threads_active == 0);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_unlock(&io_threads_mutex[j]);
-    server.io_threads_active = 1;
-}
-
-void stopThreadedIO(void) {
-    /* We may have still clients with pending reads when this function
-     * is called: handle them before stopping the threads. */
-    handleClientsWithPendingReadsUsingThreads();
-    serverAssert(server.io_threads_active == 1);
-    for (int j = 1; j < server.io_threads_num; j++)
-        pthread_mutex_lock(&io_threads_mutex[j]);
-    server.io_threads_active = 0;
-}
-
-/* This function checks if there are not enough pending clients to justify
- * taking the I/O threads active: in that case I/O threads are stopped if
- * currently active. We track the pending writes as a measure of clients
- * we need to handle in parallel, however the I/O threading is disabled
- * globally for reads as well if we have too little pending clients.
+/* This function dynamically adjusts the I/O thread number to match real-time
+ * workloads. It tracks the latest 'x' rounds of pending writes as a measure of
+ * clients needing parallel handling. If there are too few pending clients, I/O
+ * threading is disabled globally, for both reads and writes.
  *
- * The function returns 0 if the I/O threading should be used because there
- * are enough active threads, otherwise 1 is returned and the I/O threads
- * could be possibly stopped (if already active) as a side effect. */
-int stopThreadedIOIfNeeded(void) {
-    int pending = listLength(server.clients_pending_write);
-
-    /* Return ASAP if IO threads are disabled (single threaded mode). */
+ * The function returns the active IO worker thread count. */
+#define IO_THREAD_BATCH_NUM 2
+#define SMOOTH_FACTOR 0.8
```

Comment:
Yes, a hardcode number. It can take react to the target real time value, just as Ping mentioned, it decays the moving average by an order of magnitude every 10 rounds.

![image](https://github.com/valkey-io/valkey/assets/698621/7adfaa7a-497b-435a-afc5-c4f55ca68a9f)